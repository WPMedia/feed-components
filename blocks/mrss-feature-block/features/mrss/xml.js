import PropTypes from 'fusion:prop-types'
import Consumer from 'fusion:consumer'
import get from 'lodash/get'
import moment from 'moment'
import getProperties from 'fusion:properties'
import { resizerKey } from 'fusion:environment'
import { BuildContent } from '@wpmedia/feeds-content-elements'
import { generatePropsForFeed } from '@wpmedia/feeds-prop-types'
import { buildResizerURL } from '@wpmedia/feeds-resizer'
import { findVideo } from '@wpmedia/feeds-find-video-stream'

const jmespath = require('jmespath')

const rssTemplate = (
  elements,
  {
    channelTitle,
    channelDescription,
    channelPath,
    channelCopyright,
    channelTTL,
    channelUpdatePeriod,
    channelUpdateFrequency,
    channelCategory,
    channelLogo,
    imageTitle,
    imageCaption,
    imageCredits,
    itemTitle,
    itemDescription,
    pubDate,
    itemCredits,
    itemCategory,
    includeContent,
    resizerURL,
    domain,
    feedTitle,
    feedLanguage,
    videoInfo,
    mrssBuildContent,
  },
) => ({
  rss: {
    '@xmlns:atom': 'http://www.w3.org/2005/Atom',
    '@version': '2.0',
    '@xmlns:media': 'http://search.yahoo.com/mrss/',
    channel: {
      title: `${channelTitle || feedTitle + ' Videos'}`,
      link: domain,
      'atom:link': {
        '@href': `${domain}${channelPath}`,
        '@rel': 'self',
        '@type': 'application/rss+xml',
      },
      description: channelDescription || feedTitle,
      pubDate: moment.utc(Date.now()).format('ddd, DD MMM YYYY HH:mm:ss ZZ'),
      ...(channelLogo && {
        image: {
          url: buildResizerURL(channelLogo, resizerKey, resizerURL),
          title: `${channelTitle || feedTitle + ' Videos'}`,
          link: domain,
        },
      }),

      item: elements.map((s) => {
        const url = `${domain}${s.website_url || s.canonical_url}`
        const img =
          s.promo_items && (s.promo_items.basic || s.promo_items.lead_art)
        const videoStream = findVideo(s, videoInfo)

        return {
          title: `${jmespath.search(s, itemTitle)}`,
          link: url,
          ...(itemCredits &&
            (jmespath.search(img, itemCredits) || []).length && {
              'media:credit': {
                '@role': 'producer',
                $: jmespath.search(img, itemCredits).join(','),
              },
            }),
          ...(itemDescription && {
            description: jmespath.search(s, itemDescription),
          }),
          guid: {
            '@isPermaLink': false,
            '#': url,
          },
          referenceid: s._id,
          pubDate: moment
            .utc(s[pubDate])
            .format('ddd, DD MMM YYYY HH:mm:ss ZZ'),

          'media:content': {
            '@': {
              isDefault: 'true',
              ...(s.duration && {
                duration: Math.trunc(s.duration / 1000),
              }),
              ...(videoStream && {
                ...(videoInfo.url && { url: videoInfo.url }),
                ...(videoInfo.height && { height: videoInfo.height }),
                ...(videoInfo.width && { width: videoInfo.width }),
                ...(videoInfo.bitrate && { bitrate: videoInfo.bitrate }),
                ...(videoInfo.stream_type && {
                  type: videoInfo.stream_type,
                  '#': videoStream,
                }),
              }),
            },
          },
          'media:keywords': (
            jmespath.search(s, 'taxonomy.seo_keywords[*]') || []
          ).join(','),
          ...(s.description ||
            (s.subheadlines &&
              s.subheadlines.basic && {
                'media:caption': {
                  $: s.description || s.subheadlines.basic,
                },
              })),
          'media:transcript': s.transcript,
          ...(s.taxonomy &&
            s.taxonomy.primary_section &&
            s.taxonomy.primary_section.name && {
              'media:category': s.taxonomy.primary_section.name,
            }),
          ...(img &&
            img.url && {
              'media:thumbnail': {
                '@url': buildResizerURL(img.url, resizerKey, resizerURL),
              },
            }),
        }
      }),
    },
  },
})

export function Mrss({ globalContent, customFields, arcSite }) {
  const {
    resizerURL = '',
    feedDomainURL = '',
    feedTitle = '',
    feedLanguage = '',
  } = getProperties(arcSite)

  function MrssBuildContent() {
    BuildContent.call(this)
  }
  const mrssBuildContent = new MrssBuildContent()

  // can't return null for xml return type, must return valid xml template
  return rssTemplate(get(globalContent, 'content_elements', []), {
    ...customFields,
    resizerURL,
    domain: feedDomainURL,
    feedTitle,
    feedLanguage,
    mrssBuildContent,
  })
}

Mrss.propTypes = {
  customFields: PropTypes.shape({
    channelPath: PropTypes.string.tag({
      label: 'Path',
      group: 'Channel',
      description:
        'Path to the feed, excluding the domain, defaults to /arcio/mrss',
      defaultValue: '/arcio/mrss/',
    }),
    selectVideo: PropTypes.kvp.tag({
      label: 'Select video using',
      group: 'Video',
      description:
        'This criteria is used to filter videos encoded in the streams array',
      defaultValue: {
        height: '',
        width: '',
        bitrate: 5400,
        stream_type: 'mp4',
      },
    }),
    ...generatePropsForFeed('rss', PropTypes),
  }),
}
Mrss.label = 'MRSS'
export default Consumer(Mrss)
