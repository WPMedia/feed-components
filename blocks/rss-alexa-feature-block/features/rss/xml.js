import PropTypes from 'fusion:prop-types'
import Consumer from 'fusion:consumer'
import get from 'lodash/get'
import moment from 'moment'
import getProperties from 'fusion:properties'
import { resizerKey } from 'fusion:environment'
import { BuildContent } from '@wpmedia/feeds-content-elements'
import { generatePropsForFeed } from '@wpmedia/feeds-prop-types'
import { buildResizerURL } from '@wpmedia/feeds-resizer'

const jmespath = require('jmespath')

const cheerio = require('cheerio')

const getURLType = (url) => {
  const imgRegex = new RegExp(/(jpe?g|png|webp|PNG|JPG|WEBP)$/);
  const audioRegex = new RegExp('^(https?|ftp|file):\/\/(www.)?(.*?)\.(mp3)$');
  const videoRegex = new RegExp('^(https?|ftp|file):\/\/(www.)?(.*?)\.(mp4|ts)$');
  const uri = url.split(".");
  const type = uri[uri.length -1];
  const defaultType =  {label: 'Audio',type: `audio/${type}`};
  if(imgRegex.test(url)) {
    return {label: 'image',type: `image/${type}`};
  }
  if(audioRegex.test(url)){
    return {label: 'Audio',type: `audio/${type}`};
  }
  if(videoRegex.test(url)){
    return {label: 'Video',type: `video/${type}`};
  }
  
  return defaultType;
}

const rssTemplate = (
  elements,
  {
    audioAvailable,
    channelTitle,
    channelDescription,
    channelCopyright,
    channelTTL,
    channelCategory,
    channelLogo,
    itemTitle,
    pubDate,
    itemCategory,
    includeContent,
    resizerURL,
    resizerWidth,
    resizerHeight,
    domain,
    feedTitle,
    feedLanguage,
    rssBuildContent,
  },
) => ({
  rss: {
    '@version': '2.0',
    channel: {
      ...(channelTitle && { title: channelTitle }),
      link: `${domain}`,
      ...(channelDescription && { description: channelDescription }),
      lastBuildDate: moment
        .utc(new Date())
        .format('ddd, DD MMM YYYY HH:mm:ss ZZ'),
      ...(feedLanguage && { language: feedLanguage }),
      ...(channelCategory && { category: channelCategory }),
      ...(channelCopyright && {
        copyright: channelCopyright,
      }), // TODO Add default logic
      ...(channelTTL && { ttl: channelTTL }),

      ...(channelLogo && {
        image: {
          url: buildResizerURL(channelLogo, resizerKey, resizerURL),
          title: channelTitle || feedTitle,
          link: domain,
        },
      }),

      item: elements.map((s) => {
        let body, category,
        enclosureurl = jmespath.search(s, audioAvailable);
        const url = `${domain}${s.website_url || s.canonical_url || ''}`
        return {
          title: jmespath.search(s, itemTitle) || '',
          link: url,
          guid: s._id,
          pubDate: s[pubDate],
          ...(enclosureurl  &&
          {
            enclosure: {
              '@url': enclosureurl,
              '@label':  getURLType(enclosureurl).label,
              '@group': getURLType(enclosureurl).label,
              '@type': getURLType(enclosureurl).type
            }
          }),
          ...(itemCategory &&
            (category = jmespath.search(s, itemCategory)) &&
            category && { category: category }),
            description: cheerio
            .load(
              rssBuildContent.parse(
                s.content_elements || [],
                includeContent,
                domain,
                resizerKey,
                resizerURL,
                resizerWidth,
                resizerHeight,
              ) || '',
            )
            .text()
        }
      }),
    },
  },
})

export function Rss({ globalContent, customFields, arcSite }) {
  const {
    resizerURL = '',
    feedDomainURL = '',
    feedTitle = '',
    feedLanguage = '',
  } = getProperties(arcSite)

  const { width = 0, height = 0 } = customFields.resizerKVP || {};
  

  const rssBuildContent = new BuildContent()
 
  // can't return null for xml return type, must return valid xml template
  return rssTemplate(get(globalContent, 'content_elements', []), {
    ...customFields,
    resizerURL,
    resizerWidth: width,
    resizerHeight: height,
    domain: feedDomainURL,
    feedTitle,
    feedLanguage,
    rssBuildContent,
  })
}

Rss.propTypes = {
  customFields: PropTypes.shape({
    audioAvailable: PropTypes.string.tag({
      label: 'Audio',
      group: 'Audio',
      description: 'description',
      defaultValue: `content_elements[?type=='audio'].streams[].url|[0]`,
    }),
    ...generatePropsForFeed('rss', PropTypes, ['featuredImage', 'includePromo']),
  }),
}
Rss.label = 'RSS Alexa'
export default Consumer(Rss)
