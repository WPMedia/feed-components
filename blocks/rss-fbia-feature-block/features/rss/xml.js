/* eslint-disable camelcase */
import PropTypes from 'fusion:prop-types'
import Consumer from 'fusion:consumer'
import moment from 'moment'
import getProperties from 'fusion:properties'
import { resizerKey } from 'fusion:environment'
import { BuildContent } from '@wpmedia/feeds-content-elements'
import { BuildPromoItems } from '@wpmedia/feeds-promo-items'
import { generatePropsForFeed } from '@wpmedia/feeds-prop-types'
import { buildResizerURL } from '@wpmedia/feeds-resizer'
import { convert, fragment } from 'xmlbuilder2'
import URL from 'url'
const jmespath = require('jmespath')

const rssTemplate = (
  elements,
  {
    channelTitle,
    channelDescription,
    channelCopyright,
    channelTTL,
    channelUpdatePeriod,
    channelUpdateFrequency,
    channelCategory,
    channelLogo,
    imageTitle,
    imageCaption,
    imageCredits,
    itemTitle,
    itemDescription,
    itemCredits,
    pubDate,
    itemCategory,
    includePromo,
    includeContent,
    videoSelect,
    requestPath,
    resizerURL,
    resizerWidth,
    resizerHeight,
    domain,
    feedTitle,
    feedLanguage,
    promoItemsJmespath = 'promo_items.basic || promo_items.lead_art',
    PromoItems,
    fbiaBuildContent,
  },
) => ({
  rss: {
    '@xmlns:atom': 'http://www.w3.org/2005/Atom',
    '@xmlns:content': 'http://purl.org/rss/1.0/modules/content/',
    ...(itemCredits && {
      '@xmlns:dc': 'http://purl.org/dc/elements/1.1/',
    }),
    ...(channelUpdatePeriod &&
      channelUpdatePeriod !== 'Exclude field' && {
        '@xmlns:sy': 'http://purl.org/rss/1.0/modules/syndication/',
      }),
    '@version': '2.0',
    ...(includePromo && {
      '@xmlns:media': 'http://search.yahoo.com/mrss/',
    }),
    channel: {
      title: { $: channelTitle || feedTitle },
      link: `${domain}`,
      'atom:link': {
        '@href': `${domain}${requestPath}`,
        '@rel': 'self',
        '@type': 'application/rss+xml',
      },
      description: { $: channelDescription || `${feedTitle} News Feed` },
      lastBuildDate: moment
        .utc(new Date())
        .format('ddd, DD MMM YYYY HH:mm:ss ZZ'),
      ...(feedLanguage && { language: feedLanguage }),
      ...(channelCategory && { category: channelCategory }),
      ...(channelCopyright && {
        copyright: channelCopyright,
      }), // TODO Add default logic
      ...(channelTTL && { ttl: channelTTL }),
      ...(channelUpdatePeriod &&
        channelUpdatePeriod !== 'Exclude field' && {
          'sy:updatePeriod': channelUpdatePeriod,
        }),
      ...(channelUpdateFrequency &&
        channelUpdatePeriod !== 'Exclude field' && {
          'sy:updateFrequency': channelUpdateFrequency,
        }),
      ...(channelLogo && {
        image: {
          url: buildResizerURL(channelLogo, resizerKey, resizerURL),
          title: channelTitle || feedTitle,
          link: domain,
        },
      }),

      item: elements.map((s) => {
        let author, body, category
        const url = `${domain}${s.website_url || s.canonical_url}`
        const img = PromoItems.mediaTag({
          ans: s,
          promoItemsJmespath,
          resizerKey,
          resizerURL,
          resizerWidth,
          resizerHeight,
          imageTitle,
          imageCaption,
          imageCredits,
          videoSelect,
        })
        return {
          ...(itemTitle && {
            title: { $: jmespath.search(s, itemTitle) || '' },
          }),
          link: url,
          guid: {
            '#': url,
            '@isPermaLink': true,
          },
          ...(itemCredits &&
            (author = jmespath.search(s, itemCredits) || []) &&
            author.length && {
              'dc:creator': { $: author.join(', ') },
            }),
          ...(itemDescription && {
            description: { $: jmespath.search(s, itemDescription) || '' },
          }),
          pubDate: moment
            .utc(s[pubDate])
            .format('ddd, DD MMM YYYY HH:mm:ss ZZ'),
          ...(itemCategory &&
            (category = jmespath.search(s, itemCategory)) &&
            category && { category: category }),
          ...(includeContent !== 0 &&
            (body = fbiaBuildContent.buildFBContent(
              s,
              includeContent,
              domain,
              resizerWidth,
              resizerHeight,
              itemCredits,
              videoSelect,
            )) &&
            body && {
              'content:encoded': {
                $: body,
              },
            }),
          ...(includePromo &&
            img && {
              '#': img,
            }),
        }
      }),
    },
  },
})

export function FbiaRss({ globalContent, customFields, arcSite, requestUri }) {
  const {
    resizerURL = '',
    feedDomainURL = 'http://localhost.com',
    feedTitle = '',
    feedLanguage = '',
  } = getProperties(arcSite)
  const { width = 0, height = 0 } = customFields.resizerKVP || {}
  const requestPath = new URL.URL(requestUri, feedDomainURL).pathname
  let metaTags
  try {
    metaTags =
      customFields.metaTags &&
      convert(customFields.metaTags, { format: 'object' })
  } catch {
    metaTags = ''
  }

  function FbiaBuildContent({
    itemTitle,
    itemDescription,
    articleStyle,
    likesAndComments,
    metaTags,
    adPlacement,
    adDensity,
    placementSection,
    iframeHxW = {},
    raw_html_processing = 'exclulde',
  }) {
    BuildContent.call(this)

    this.buildHTMLHead = (s, domain, resizerWidth, resizerHeight) => {
      const img =
        s.promo_items && (s.promo_items.basic || s.promo_items.lead_art)
      const url = `${domain}${s.website_url || s.canonical_url || ''}`
      const title = (itemTitle && jmespath.search(s, itemTitle)) || ''
      return {
        link: {
          '@rel': 'canonical',
          '@href': url,
        },
        title: title,
        meta: [
          {
            '@property': 'og:title',
            '@content': title,
          },
          {
            '@property': 'og:url',
            '@content': url,
          },
          {
            '@property': 'og:description',
            '@content': jmespath.search(s, itemDescription || '_blank') || '',
          },
          {
            '@property': 'fb:use_automatic_ad_placement',
            '@content': adPlacement.toLowerCase().startsWith('enable')
              ? `enable=true ad_density=${adDensity || 'default'}`
              : 'enable=false',
          },
          {
            // The version of Instant Articles markup format being used by this article.
            '@property': 'op:markup_version',
            '@content': 'v1.0',
          },
          {
            '@property': 'fb:article_style',
            '@content': articleStyle || 'default',
          },
          {
            '@property': 'og:image',
            '@content':
              (img && buildResizerURL(img.url, resizerKey, resizerURL)) || '',
          },
          {
            '@property': 'fb:likes_and_comments',
            '@content': likesAndComments || 'disable',
          },
        ],
        ...(metaTags && { '#': metaTags }),
      }
    }
    this.buildHTMLBody = (
      s,
      numRows,
      domain,
      resizerWidth,
      resizerHeight,
      itemCredits,
    ) => {
      const authorDescription = (
        jmespath.search(s, 'credits.by[].description') || []
      ).filter((i) => i)
      const lastUpdatedDate = jmespath.search(s, 'last_updated_date')
      const image =
        s.promo_items && (s.promo_items.basic || s.promo_items.lead_art)
      const primarySite =
        jmespath.search(s, 'taxonomy.primary_section.name') || ''
      const header = []
      let description, author
      const adScripts = customFields.adScripts || ''
      if (placementSection)
        header.push({
          section: {
            '@class': 'op-ad-template',
            '#': [placementSection],
          },
        })

      header.push({ h1: jmespath.search(s, itemTitle || '_blank') || '' })

      if (
        itemDescription &&
        (description = jmespath.search(s, itemDescription)) &&
        description
      )
        header.push({ h2: description })

      header.push({
        time: [
          {
            '@': {
              datetime: lastUpdatedDate,
              class: 'op-modified',
            },
            '#': lastUpdatedDate,
          },
          {
            '@datetime': s[customFields.pubDate],
            '@class': 'op-published',
            '#': s[customFields.pubDate],
          },
        ],
      })

      if (
        itemCredits &&
        (author = jmespath.search(s, itemCredits) || []) &&
        author.length
      )
        header.push({
          address: {
            // a list of authors
            a: author.map((s) => s.toUpperCase()),
          },
        })
      if (customFields.includePromo && image && image.url)
        header.push({
          figure: {
            '@class': 'fb-feed-cover',
            img: {
              '@src': buildResizerURL(
                image.url,
                resizerKey,
                resizerURL,
                resizerWidth,
                resizerHeight,
              ),
            },
            ...(customFields.imageCaption &&
              jmespath.search(image, customFields.imageCaption) && {
                figcaption: {
                  '@class': 'op-vertical-below op-small',
                  '#': `${jmespath.search(image, customFields.imageCaption)}`,
                  ...((
                    jmespath.search(
                      image,
                      customFields.imageCredits || '_blank',
                    ) || []
                  ).length && {
                    cite: {
                      '@class': 'op-small',
                      '#': jmespath
                        .search(image, customFields.imageCredits)
                        .join(','),
                    },
                  }),
                },
              }),
          },
        })
      if (primarySite.length)
        header.push({
          h2: {
            '@class': 'op-kicker',
            '#': primarySite,
          },
        })
      return {
        article: {
          header: {
            '#': header,
          },
          '#': ['<tHe_BoDy_GoEs_HeRe/>', ...(adScripts && [adScripts])],
          footer: {
            '#': {
              ...(authorDescription.length && {
                aside: authorDescription.join(','),
              }),
              small:
                customFields.channelCopyright ||
                '© ' + moment.utc(new Date()).year() + ' ' + feedTitle,
            },
          },
        },
      }
    }

    this.image = (
      element,
      resizerKey,
      resizerURL,
      resizerWidth,
      resizerHeight,
    ) => {
      const credits = jmespath.search(element, 'credits.by[].name') || []
      return {
        figure: {
          img: {
            '@': {
              src: buildResizerURL(
                element.url,
                resizerKey,
                resizerURL,
                resizerWidth,
                resizerHeight,
              ),
            },
          },
          ...(element.caption && {
            figcaption: {
              '@class': 'op-vertical-below op-small',
              '#': {
                '#': element.caption,
                ...(credits.length && {
                  cite: {
                    '@class': 'op-small',
                    '#': '(' + credits.join(',') + ')',
                  },
                }),
              },
            },
          }),
        },
      }
    }
    this.text = (element) => {
      // handle text, raw_html
      // all have a string in element.content
      // this is also used by buildContentQuote
      let item
      const { width = 0, height = 0 } = iframeHxW
      if (element.content && typeof element.content === 'string') {
        if (element.type === 'raw_html') {
          switch (raw_html_processing) {
            case 'wrap':
              item = {
                figure: {
                  '@class': 'op-interactive',
                  iframe: {
                    ...(width && { '@width': width }),
                    ...(height && { '@height': height }),
                    '#': element.content,
                  },
                },
              }
              break
            case 'include':
              item = {
                '#': element.content,
              }
              break
          }
        } else {
          item = {
            p: {
              '@id': element._id,
              '#': element.content,
            },
          }
        }
      }
      return item
    }
    // noinspection SpellCheckingInspection
    this.oembed = (element) => {
      const embed = element.raw_oembed.html // wrap in <figure class="op-interactive">
      const { width = 0, height = 0 } = iframeHxW

      return {
        figure: {
          '@class': 'op-interactive',
          iframe: {
            ...(width && { '@width': width }),
            ...(height && { '@height': height }),
            '#': embed,
          },
        },
      }
    }
    this.header = (element) => {
      let item
      if (element.content && typeof element.content === 'string') {
        item = {}
        item[`h${element.level ? (element.level < 2 ? 1 : 2) : 1}`] =
          element.content
      }
      return item
    }

    this.table = (element) => {
      const header = []
      const rows = []
      let row

      element.header &&
        element.header.forEach((headerItem) => {
          header.push(headerItem.content)
        })

      element.rows &&
        element.rows.forEach((tableRows) => {
          row = []
          tableRows.forEach((rowItem) => {
            row.push(rowItem.content)
          })
          rows.push({ tr: { td: row } })
        })

      return {
        figure: {
          table: {
            thead: { tr: { th: header } },
            tbody: { '#': rows },
          },
        },
      }
    }

    this.buildFBContent = (
      s,
      numRows,
      domain,
      resizerWidth,
      resizerHeight,
      itemCredits,
      videoSelect,
    ) => {
      const fbiaContent = {
        html: {
          '@lang': feedLanguage,
          head: this.buildHTMLHead(s, domain, resizerWidth, resizerHeight),
          body: this.buildHTMLBody(
            s,
            numRows,
            domain,
            resizerWidth,
            resizerHeight,
            itemCredits,
            videoSelect,
          ),
        },
      }
      // breaking these up because I'm
      // seeing xml validation fail on <br> without a closing slash
      const htmlBody = '<!doctype html>'.concat(
        fragment(fbiaContent).toString(),
      )
      const parsedBody = this.parse(
        s.content_elements ?? [],
        numRows,
        domain,
        resizerKey,
        resizerURL,
        resizerWidth,
        resizerHeight,
        videoSelect,
      )
      return htmlBody.replace('<tHe_BoDy_GoEs_HeRe/>', parsedBody)
    }
  }

  const fbiaBuildContent = new FbiaBuildContent({
    ...customFields,
    metaTags,
  })

  const PromoItems = new BuildPromoItems()

  // can't return null for xml return type, must return valid xml template
  return rssTemplate(globalContent.content_elements || [], {
    ...customFields,
    requestPath,
    resizerURL,
    resizerWidth: width,
    resizerHeight: height,
    domain: feedDomainURL,
    feedTitle,
    feedLanguage,
    fbiaBuildContent,
    PromoItems,
  })
}
// Reference for fb options: https://developers.facebook.com/docs/instant-articles/reference/article/
FbiaRss.propTypes = {
  customFields: PropTypes.shape({
    articleStyle: PropTypes.string.tag({
      label: 'Article Style',
      group: 'Facebook Options',
      description:
        'This parameter is optional and your default style is applied to this article if you do not specify an article style in your markup',
      defaultValue: 'default',
    }),
    likesAndComments: PropTypes.oneOf(['enable', 'disable']).tag({
      defaultValue: 'disable',
      name: 'Likes and Comments',
      group: 'Facebook Options',
      description: 'Enable or disable likes and comments on the article',
    }),
    metaTags: PropTypes.string.tag({
      label: 'Additional meta tags',
      group: 'Facebook Options',
      description:
        'Enter additonal meta tags here in the format <meta property="prop" content="content"/> ',
      defaultValue: '',
    }),
    adPlacement: PropTypes.oneOf(['enable', 'disable']).tag({
      name: 'Auto Ad Placement',
      group: 'Facebook Options',
      description:
        'Enables automatic placement of ads within this article. Only enable this option if you configure an ad.',
      defaultValue: 'disable',
    }),
    adDensity: PropTypes.oneOf(['default', 'medium', 'low']).tag({
      name: 'Ad Density',
      group: 'Facebook Options',
      description:
        'How frequently you would like ads to appear in your article: default (<250 word gap), medium (350 word gap), low (>450 word gap)',
      defaultValue: 'default',
    }),
    placementSection: PropTypes.string.tag({
      label: 'Facebook Ad',
      group: 'Facebook Options',
      description:
        'Enter Javascript that goes between <section class="op-ad-template"></section> in beginning of the body\'s header for recirculation ads that come from Facebook advertisers; leave blank if not used.',
      defaultValue: '',
    }),
    adScripts: PropTypes.string.tag({
      label: 'Analytic Scripts',
      group: 'Facebook Options',
      description:
        'Javascript wrapped in the <figure class=‘op-tracker’> tag can be added to the article for ads and analytics. Multiple scripts can be included, usually each in the own iframe',
      defaultValue: '',
    }),
    iframeHxW: PropTypes.kvp.tag({
      label: 'oembed iframe height and width',
      group: 'Facebook Options',
      description: 'Height and/or width to use in oembed iframes',
      defaultValue: {
        width: 0,
        height: 0,
      },
    }),
    raw_html_processing: PropTypes.oneOf(['exclude', 'include', 'wrap']).tag({
      label: {
        exclude: 'raw_html elements',
        include: 'raw_html elements',
        wrap: 'wrap in <figure class="op-interactive"> <iframe>',
      },
      group: 'Facebook Options',
      description:
        'Should raw_html elements be excluded, included or wrapped in <figure class="op-interactive"> <iframe> tags. default exclude',
      defaultValue: 'exclude',
    }),
    ...generatePropsForFeed('rss', PropTypes),
  }),
}

FbiaRss.label = 'RSS FBIA'
export default Consumer(FbiaRss)
