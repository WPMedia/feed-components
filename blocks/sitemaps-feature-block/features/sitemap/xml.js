import PropTypes from 'prop-types'
import Consumer from 'fusion:consumer'
import get from 'lodash/get'
import getProperties from 'fusion:properties';
import {resizerKey} from 'fusion:environment'

export class Sitemap {
  constructor(props) {
    this.props = props
    const { arcSite } = props
    this.resizerURL = getProperties(arcSite).resizerURL
    this.feedDomainURL = getProperties(arcSite).feedDomainURL
    this.fetchContent({
      site: {
        source: 'feeds-site-service',
        query: {
          website: arcSite,
        },
      },
    })
  }

  getImgURL(element)  {
    const buildURL = _url => {
      if (typeof window === 'undefined') {
        const Thumbor = require('thumbor-lite')
        const thumbor = new Thumbor(resizerKey, this.resizerURL)
        let imgSrc = _url.replace(/^http[s]?:\/\//, '').replace(' ', '%20')
        if (imgSrc.includes('?')) {
          imgSrc = imgSrc.replace('?', '%3F')
        }

        return thumbor
          .setImagePath(imgSrc)
          .resize(1200, 630)
          .buildUrl()
      }
      return null
    }

    if (
      element.promo_items &&
      element.promo_items.basic &&
      element.promo_items.basic.url
    ) {
      return buildURL(element.promo_items.basic.url)
    }

    return ''
  }

  sitemapTemplate  (
    elements,
    { domain, lastMod, changeFreq, includePromo, priority, imageTitle, imageCaption }

  )  {
    return {
      urlset: {
        '@xmlns': 'http://www.sitemaps.org/schemas/sitemap/0.9',
        ...(includePromo && {
          '@xmlns:image': 'http://www.google.com/schemas/sitemap-image/1.1',
        }),
        url: elements.map(s => {
          const img =
            s.promo_items && (s.promo_items.basic || s.promo_items.lead_art)
          return {
            loc: `${domain}${s.canonical_url}`,
            ...({ lastmod: s[lastMod] }),
            ...(changeFreq !== 'Exclude from sitemap' && { changefeq: changeFreq }),
            ...(priority !== 'Exclude from sitemap' && { priority: priority }),
            ...(includePromo &&
              img && {
                'image:image': {
                  'image:loc': this.getImgURL(s),
                  ...(img[imageCaption] && {
                    'image:caption': { $: img[imageCaption] },
                  }),
                  ...(img[imageTitle] && {
                    'image:title': { $: img[imageTitle] },
                  }),
                },
              }),
          }
        }),
      },
    }
  }

  render() {
    const { site } = this.state || {}
    const { globalContent, customFields } = this.props

    // can't return null for xml return type, must return valid xml template
    if(!site) {
      return this.sitemapTemplate([], customFields)
    }
    const elements = get(globalContent, 'content_elements', [])
    return this.sitemapTemplate(elements, { ...customFields, domain: this.feedDomainURL})
  }

}

Sitemap.propTypes = {
  customFields: PropTypes.shape({
    imageTitle: PropTypes.string.tag({
      label: 'ANS image title key',
      group: 'Field Mapping',
      description:
        'ANS value for associated story used for the <image:title> sitemap tag',
      defaultValue: 'title',
    }),
    imageCaption: PropTypes.string.tag({
      label: 'ANS image title key',
      group: 'Field Mapping',
      description:
        'ANS value for associated story image used for the <image:caption> sitemap tag',
      defaultValue: 'caption',
    }),
    lastMod: PropTypes.oneOf([
      'created_date', 'display_date', 'first_publish_date', 'last_updated_date', 'publish_date'
    ]).tag({
      label: 'Last Modified Date',
      group: 'Format',
      description: 'Which date field should be used in the sitemap',
      defaultValue: 'last_updated_date',
    }),
    changeFreq: PropTypes.oneOf([
      'always', 'hourly', 'daily', 'weekly', 'monthly', 'yearly', 'never', 'Exclude from sitemap'
    ]).tag({
      label: 'change frequency',
      group: 'Format',
      description: 'What is the Change frequency of the sitemap',
      defaultValue: 'always',
    }),
    includePromo: PropTypes.boolean.tag({
      label: 'Include promo images?',
      group: 'Format',
      description: 'Include an image in the sitemap',
      defaultValue: true,
    }),
    priority: PropTypes.oneOf([
        '0.0', '0.1', '0.2', '0.3', '0.4', '0.5', '0.6', '0.7', '0.8', '0.9', '1.0', 'Exclude from sitemap'
    ]).tag({
      label: 'priority',
      group: 'Format',
      description: 'What is the priority of the sitemap',
      defaultValue: '0.5',
    }),
  }),
}
Sitemap.label = 'Standard Sitemap'
export default Consumer(Sitemap)
