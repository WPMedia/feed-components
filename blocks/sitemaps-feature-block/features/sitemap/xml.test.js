// eslint-disable-next-line no-unused-vars
import Consumer from 'fusion:consumer'
import { Sitemap } from './xml'

jest.mock('thumbor-lite')
it('returns template with default values', () => {
  const sitemap = Sitemap({
    arcSite: 'the-globe',
    globalContent: {
      content_elements: [
        {
          last_updated_date: '2020-04-07T15:02:08.918Z',
          website_url: '/food/2020/04/07/tips-for-safe-hand-washing',
          promo_items: {
            basic: {
              title: 'Hand Washing',
              url:
                'https://arc-anglerfish-arc2-prod-demo.s3.amazonaws.com/public/JTWX7EUOLJE4FCHYGN2COQAERY.png',
            },
          },
        },
      ],
    },
    customFields: {
      priority: '0.5',
      includePromo: true,
      changeFreq: 'always',
      lastMod: 'last_updated_date',
      imageTitle: 'title',
      imageCaption: 'caption',
    },
  })
  expect(sitemap).toMatchSnapshot()
})
