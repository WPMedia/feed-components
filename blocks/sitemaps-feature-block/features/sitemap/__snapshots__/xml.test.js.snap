// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`returns template with default values 1`] = `
Object {
  "urlset": Object {
    "@xmlns": "http://www.sitemaps.org/schemas/sitemap/0.9",
    "@xmlns:image": "http://www.google.com/schemas/sitemap-image/1.1",
    "url": Array [
      Object {
        "changefreq": "always",
        "image:image": Object {
          "image:loc": null,
          "image:title": Object {
            "$": "Hand Washing",
          },
        },
        "lastmod": "2020-04-07T15:02:08.918Z",
        "loc": "http://demo-prod.origin.arcpublishing.com/food/2020/04/07/tips-for-safe-hand-washing",
        "priority": "0.5",
      },
    ],
  },
}
`;
